specifyweb/frontend/js_src/lib/components/WorkBench/WbView.tsx

```diff
@@ -50,12 +50,552 @@ import { parseWbMappings } from './mapping';
 import { fetchWbPickLists } from './pickLists';
 import { WbUploaded } from './Results';
 import { wbViewTemplate } from './Template';
-import { WbActions } from './WbActions';
+import { WbActions, WbActionsReact } from './WbActions';
 import { WbUtils } from './WbUtils';
 import { WbValidation } from './WbValidation';
+import type { LocalizedString } from 'typesafe-i18n';
+import { className } from '../Atoms/className';
+import { Input } from '../Atoms/Form';
+import { DataSetName } from './DataSetMeta';
+import { WbSpreadsheet } from './WbSpreadsheet';
+import { useBooleanState } from '../../hooks/useBooleanState';
+import { WbToolkit } from './WbToolkit';
+import { getCache } from '../../utils/cache';
+import { WbAdvancedSearch } from './AdvancedSearch';
+import type { WbSearchPreferences } from './AdvancedSearch';
+import { RollbackConfirmation } from './Components';
+import { WbStatus as WbStatusComponent } from './Status';
+import type { Status } from '../WbPlanView/Wrapped';
+import { loadingBar } from '../Molecules';
+import { CreateRecordSetButton } from './RecordSet';
 
 export type WbStatus = 'unupload' | 'upload' | 'validate';
 
+function Navigation({
+  name,
+  label,
+}: {
+  readonly name: string;
+  readonly label: LocalizedString;
+}): JSX.Element {
+  return (
+    <span
+      aria-atomic
+      className="wb-navigation-section flex rounded"
+      data-navigation-type={name}
+    >
+      <Button.Small
+        className="wb-cell-navigation brightness-80 hover:brightness-70 p-2 ring-0"
+        data-navigation-direction="previous"
+        variant="bg-inherit text-gray-800 dark:text-gray-100"
+        onClick={f.never}
+      >
+        {'<'}
+      </Button.Small>
+      <Button.Small
+        className={`
+          wb-navigation-text hover:brightness-70 grid grid-cols-[auto_1fr_auto_1fr_auto]
+          items-center ring-0
+          ${className.ariaHandled}
+        `}
+        title={wbText.clickToToggle()}
+        variant="bg-inherit text-gray-800 dark:text-gray-100"
+        onClick={f.never}
+      >
+        {label} (<span className="wb-navigation-position text-center">0</span>/
+        <span className="wb-navigation-total">0</span>)
+      </Button.Small>
+      <Button.Small
+        className="wb-cell-navigation brightness-80 hover:brightness-70 p-2 ring-0"
+        data-navigation-direction="next"
+        type="button"
+        variant="bg-inherit text-gray-800 dark:text-gray-100"
+        onClick={f.never}
+      >
+        {'>'}
+      </Button.Small>
+    </span>
+  );
+}
+
+export type DialogHandlers = {
+  show: boolean;
+  open: () => void;
+  close: () => void;
+  toggle: () => void;
+};
+
+export type HandlersObject = {
+  [key: string]: DialogHandlers;
+};
+
+// Returns an object consisting of handler objects for dialog actions in the Workbench
+function useWbViewHandlers() {
```

Comment:
I would encourage you to try having the smallest states possible, rather than one large state.

For example, there could be a LeafletMap component, that would just render the button, and call useBooleanState inside. then, when button is clicked, this component would render a dialog with the map.

This way, the state is defined very close to where it's used (making the application more readable and maintainable), but also performance is improved (because if the top-level state changes, all workbench components are re-rendered. if a local state is re-rendered, parent's don't need to be touched)

Reply:
Thank you. I'll refactor this once I iron out other bugs